### PR TITLE
[Merged by Bors] - TLS Support for MQTT connector.

### LINKED ISSUE #164: mqtt - Port from the URL is not being respected.
Port is fixed to 1883, should use port from URL

[MqttOptions::new](https://docs.rs/rumqttc/latest/rumqttc/struct.MqttOptions.html#method.new)
```rust
MqttOptions::new(client_id, host, port)
```

src/main.rs
```rust
#[structopt(short, long)]
mqtt_url: String,

/// -- clip --

let mut mqttoptions = MqttOptions::new(client_id, mqtt_url, 1883);
```

### LINKED ISSUE #173: Support TLS for MQTT

### PR DESCRIPTION
**Pull request by simlay**
Closes #173.
Closes #164.

### CODE REVIEW

File: rust-connectors/sources/mqtt/src/main.rs [hunk 1]
```diff
@@ -84,11 +94,32 @@ fn main() -> Result<(), MqttConnectorError> {
             %mqtt_topic,
             %client_id
         );
+        let port = if opts.mqtt_tls { 8883 } else { 1883 };
```
Comment (morenol) [inline rust-connectors/sources/mqtt/src/main.rs:97]: I think that we should add a `mqtt_port` configuration just in case it is using a custom port.

File: rust-connectors/sources/mqtt/README.md [hunk 2]
```diff
@@ -14,12 +14,16 @@ MQTT V3.1.1 and V5
...
-| Option     | default  | type   | description |
-| :---       | :---     | :---   | :----       |
-| timeout    | 60       | u64    | mqtt broker connect timeout in seconds |
-| mqtt_url   | -        | String | mqtt broker Host |
-| mqtt_topic | -        | String | mqtt topic to subscribe and source events from |
-| client_id  | UUID V4  | String | mqtt client ID |
+| Option        | default  | type   | description                                             |
+| :---          | :---     | :---   | :----                                                   |
+| timeout       | 60       | u64    | mqtt broker connect timeout in seconds                  |
+| mqtt_url      | -        | String | mqtt broker Host                                        |
```
Comment (pinkforest) [inline rust-connectors/sources/mqtt/README.md:20]: mqtt_port, mqtt_tls should be incorporated in mqtt_url not separately perhaps? mqtt url should support differnet port / tls schemes perhaps? I know it's separate issue on #164 but?

File: rust-connectors/sources/mqtt/README.md [hunk 3]
```diff
@@ -14,12 +14,16 @@ MQTT V3.1.1 and V5
...
+| mqtt_topic    | -        | String | mqtt topic to subscribe and source events from          |
+| mqtt_tls      | -        | Flag   | Enables TLS on the MQTT connection                      |
+| mqtt_port     | -        | u16    | Port to use for the MQTT connection                     |
+| mqtt_username | -        | String | mqtt username to authenticate with                      |
+| MQTT_PASSWORD | -        | String | mqtt password to authenticate with in the secrets list  |
```
Comment (pinkforest) [inline rust-connectors/sources/mqtt/README.md:25]: these caps look mean(ish) :eyes: or is it our way of trying to differentiate "secrets" from env?

File: rust-connectors/sources/mqtt/src/main.rs [hunk 4]
```diff
@@ -84,11 +97,40 @@ fn main() -> Result<(), MqttConnectorError> {
             %mqtt_topic,
             %client_id
         );
+        let port = if let Some(port) = opts.mqtt_port {
+            port
+        } else {
+            if opts.mqtt_tls {
+                8883
+            } else {
```
Comment (pinkforest) [inline rust-connectors/sources/mqtt/src/main.rs:105]: this could be pain to deal with in forwards compatibility?

File: rust-connectors/sources/mqtt/src/main.rs [hunk 5]
```diff
@@ -84,11 +97,40 @@ fn main() -> Result<(), MqttConnectorError> {
...
         let timeout = std::time::Duration::from_secs(mqtt_timeout_seconds);
 
-        let mut mqttoptions = MqttOptions::new(client_id, mqtt_url, 1883);
+        let mut mqttoptions = MqttOptions::new(client_id, mqtt_url, port);
+        if let (Some(username), Some(password)) = (opts.mqtt_username, opts.mqtt_password) {
+            mqttoptions.set_credentials(username, password);
+        }
         mqttoptions.set_keep_alive(timeout);
+        if opts.mqtt_tls {
+            // Use rustls-native-certs to load root certificates from the operating system.
```
Comment (pinkforest) [inline rust-connectors/sources/mqtt/src/main.rs:118]: we probably need something for cert management overall perhaps coordinated across connectors. 

also probably we should perhaps also do tests to ensure we validate cert negative/positive.

File: rust-connectors/sources/mqtt/src/main.rs [hunk 6]
```diff
@@ -20,6 +21,18 @@ struct MqttOpts {
     #[structopt(short, long)]
     mqtt_url: String,
 
+    #[structopt(long)]
+    mqtt_tls: bool,
+
+    #[structopt(long)]
+    mqtt_username: Option<String>,
+
+    #[structopt(long, env = "MQTT_PASSWORD", hide_env_values = true)]
```
Comment (pinkforest) [inline rust-connectors/sources/mqtt/src/main.rs:30]: it's a nit but probably may need to hide from CLI options so people are disciplined to use ENV in dev without exposing password via CLI args.

File: rust-connectors/sources/mqtt/src/main.rs [hunk 7]
```diff
@@ -20,6 +21,18 @@ struct MqttOpts {
     #[structopt(short, long)]
     mqtt_url: String,
```
Comment (pinkforest) [inline rust-connectors/sources/mqtt/src/main.rs:22]: should we just perhaps deprecate and change it to host as to clear intent?

File: rust-connectors/sources/mqtt/README.md (hunk 2 above)
Comment (pinkforest) [reply rust-connectors/sources/mqtt/README.md:20]: or should we just perhaps deprecate and change it to host as to clear intent?

File: rust-connectors/sources/mqtt/src/main.rs [hunk 8]
```diff
@@ -84,11 +97,40 @@ fn main() -> Result<(), MqttConnectorError> {
...
         mqttoptions.set_keep_alive(timeout);
+        if opts.mqtt_tls {
+            // Use rustls-native-certs to load root certificates from the operating system.
+            let mut root_cert_store = rustls::RootCertStore::empty();
+            for cert in
+                rustls_native_certs::load_native_certs().expect("could not load platform certs")
+            {
+                root_cert_store
+                    .add(&rustls::Certificate(cert.0))
+                    .expect("Failed to load DER from platform cert");
```
Comment (pinkforest) [inline rust-connectors/sources/mqtt/src/main.rs:125]: should we perhaps add option to inject "pinned" cert pem's via metadata?

**Review by pinkforest — CHANGES_REQUESTED** [review]
few things to nit perhaps

File: rust-connectors/sources/mqtt/README.md (hunk 3 above)
Comment (simlay) [reply rust-connectors/sources/mqtt/README.md]: Yeah. I'm open to having them be it's own table.

File: rust-connectors/sources/mqtt/src/main.rs (hunk 8 above)
Comment (simlay) [reply rust-connectors/sources/mqtt/src/main.rs]: I'm not sure how we can inject pinned certs at this moment. I think it'd involve a bunch of the stuff in the fluvio-sc to copy or mount stuff inside the container.

File: rust-connectors/sources/mqtt/src/error.rs [hunk 9]
```diff
@@ -16,12 +17,16 @@ pub enum MqttConnectorError {
...
     MqttClient(#[from] MqttClientError),
+    #[error("Mqtt OptionError: `{0:#?}`.")]
+    MqttOption(#[from] MqttOptionError),
     #[error("Fluvio error: `{0:?}`.")]
     Fluvio(#[from] FluvioError),
     #[error("Clap Error: `{0:#?}`.")]
     Clap(#[from] ClapError),
     #[error("Serde Json Error: `{0:#?}`.")]
     SerdeJson(#[from] SerdeJsonError),
+    #[error("URL Parse Error: `{0:#?}`.")]
```
Comment (morenol) [inline rust-connectors/sources/mqtt/src/error.rs:28]: Do we need the debug representation here? I wonder if this could leak any secrets

Comment (simlay) [reply rust-connectors/sources/mqtt/src/error.rs:28]: Ah good catch.

File: rust-connectors/sources/mqtt/src/main.rs [hunk 10]
```diff
@@ -69,26 +70,24 @@ fn main() -> Result<(), MqttConnectorError> {
...
 
         let client_id = opts
             .client_id
             .unwrap_or_else(|| uuid::Uuid::new_v4().to_string());
 
         info!(
-            %mqtt_timeout_seconds,
-            %mqtt_url,
+            timout=&opts.timeout,
+            mqtt_url=%opts.mqtt_url,
```
Comment (morenol) [inline rust-connectors/sources/mqtt/src/main.rs:82]: maybe could leak here the password too?

**Comment by morenol** [comment]
lgtm, but when I tried to run this, it fails:

`cargo run --bin mqtt -- --fluvio-topic mqtt-data --mqtt-topic "/hfp/v2/journey/ongoing/vp/bus/0018/#" --mqtt-url mqtts://mqtt.hsl.fi:8883/`

```
2022-02-16T20:52:28.925270Z  INFO mqtt: Initializing MQTT connector
2022-02-16T20:52:28.927213Z  INFO mqtt: mqtt_url=mqtts://mqtt.hsl.fi:8883/ fluvio_topic=aaa mqtt_topic=/hfp/v2/journey/ongoing/vp/bus/0018/# client_id=7feabb61-a0a7-4236-8324-3ef26419652c
Error: MqttOption(ClientId)
```

Am I doing something wrong?

File: rust-connectors/sources/mqtt/src/error.rs [hunk 11]
```diff
@@ -16,12 +17,16 @@ pub enum MqttConnectorError {
     MqttConnection(#[from] MqttConnectionError),
     #[error("Mqtt client error: `{0:#?}`.")]
     MqttClient(#[from] MqttClientError),
+    #[error("Mqtt OptionError: `{0:#?}`.")]
```
Comment (morenol) [inline rust-connectors/sources/mqtt/src/error.rs:20]: I think that we should use the display representation in most of these errors

Comment (simlay) [reply rust-connectors/sources/mqtt/src/error.rs]: Done.

File: rust-connectors/sources/mqtt/src/main.rs [hunk 12]
```diff
@@ -93,9 +93,26 @@ fn main() -> Result<(), MqttConnectorError> {
                 %client_id
             );
         }
-        let mut mqttoptions = MqttOptions::try_from(url)?;
+        let mut mqttoptions = MqttOptions::try_from(url.clone())?;
         mqttoptions.set_keep_alive(mqtt_timeout_seconds);
+        if url.scheme() == "mqtts" {
```
Comment (morenol) [inline rust-connectors/sources/mqtt/src/main.rs:98]: This should be `mqtts` or `ssl` right?

File: rust-connectors/sources/mqtt/src/main.rs [hunk 13]
```diff
@@ -93,9 +93,26 @@ fn main() -> Result<(), MqttConnectorError> {
...
+            }
+            let client_config = ClientConfig::builder()
+                .with_safe_defaults()
+                .with_root_certificates(root_cert_store)
+                .with_no_client_auth();
+
+            mqttoptions.set_transport(Transport::tls_with_config(client_config.into()));
+        }
         let producer = opts.common.create_producer().await?;
+        info!("MQTT OPTIONS {:?}", mqttoptions);
```
Comment (morenol) [inline rust-connectors/sources/mqtt/src/main.rs:115]: Should we remove this line, right? or maybe change to `debug!`

File: Cargo.toml [hunk 14]
```diff
@@ -12,3 +12,5 @@ members = [
     "rust-connectors/utils/fluvio-smartstream-map/"
 ]
 resolver = "2"
+[patch.crates-io]
+rumqttc = { git = "https://github.com/bytebeamio/rumqtt" }
```
Comment (morenol) [inline Cargo.toml:16]: I think that we should remove Cargo.lock from .gitignore as per https://doc.rust-lang.org/cargo/guide/cargo-toml-vs-cargo-lock.html and then add that to the repo

**Review by morenol — COMMENTED** [review]
lgtm, it is working as expected now using mqtts and helsinski transit data. I added a few comments

File: rust-connectors/sources/mqtt/src/main.rs (hunk 13 above)
Comment (simlay) [reply rust-connectors/sources/mqtt/src/main.rs:115]: I didn't mean to leave it in.

File: rust-connectors/sources/mqtt/src/main.rs (hunk 12 above)
Comment (simlay) [reply rust-connectors/sources/mqtt/src/main.rs:98]: ah yes.

File: rust-connectors/sources/mqtt/src/main.rs (hunk 8 above)
Comment (pinkforest) [reply rust-connectors/sources/mqtt/src/main.rs]: The cert is just base64 string and it could be injected via metadata?

File: rust-connectors/sources/mqtt/README.md [hunk 15]
```diff
@@ -12,14 +12,20 @@ MQTT V3.1.1 and V5
...
-| timeout    | 60       | u64    | mqtt broker connect timeout in seconds |
-| mqtt_url   | -        | String | mqtt broker Host |
-| mqtt_topic | -        | String | mqtt topic to subscribe and source events from |
-| client_id  | UUID V4  | String | mqtt client ID |
+#### Paramaters
+
+| Option        | default  | type   | description                                             |
+| :---          | :---     | :---   | :----                                                   |
+| timeout       | 60       | u64    | mqtt broker connect timeout in seconds                  |
+| mqtt_url      | -        | String | mqtt broker Host                                        |
```
Comment (pinkforest) [inline rust-connectors/sources/mqtt/README.md:20]: nit: could we perhaps change the description to reflect that it's url now? thx

**Review by morenol — APPROVED** [review]
lgtm, just the nit that @pinkforest suggested
